src/dragonball/src/device_manager/vfio_dev_mgr/mod.rs

```diff
@@ -0,0 +1,805 @@
+        // create and attach pci root bus
+        #[cfg(all(feature = "hotplug", feature = "host-device"))]
+        if ctx.pci_hotplug_enabled {
+            let _ = self
+                .create_pci_manager(
+                    ctx.irq_manager.clone(),
+                    ctx.io_context.clone(),
+                    ctx.res_manager.clone(),
+                )
+                .map_err(StartMicroVmError::CreateVfioDevice)?;
+        }
+        for (idx, info) in self.info_list.clone().iter().enumerate() {
+            self.create_device(&info.config, ctx, idx)
+                .map_err(StartMicroVmError::CreateVfioDevice)?;
+        }
+        Ok(())
+    }
+
+    pub fn remove_device(&mut self, ctx: &mut DeviceOpContext, hostdev_id: &str) -> Result<()> {
+        if !cfg!(feature = "hotplug") {
+            return Err(VfioDeviceError::UpdateNotAllowedPostBoot);
+        }
+
+        slog::info!(
+            ctx.logger(),
+            "remove VFIO device";
+            "subsystem" => "vfio_dev_mgr",
+            "hostdev_id" => hostdev_id,
+        );
+        let device_index = self
+            .get_index_of_hostdev_id(hostdev_id)
+            .ok_or(VfioDeviceError::InvalidConfig)?;
+        let mut info = self
+            .info_list
+            .remove(device_index)
+            .ok_or(VfioDeviceError::InvalidConfig)?;
+
+        self.remove_vfio_device(ctx, &mut info)
+    }
+
+    /// prepare to remove device
+    pub fn prepare_remove_device(
+        &self,
+        ctx: &DeviceOpContext,
+        hostdev_id: &str,
+        result_sender: Sender<Option<i32>>,
+    ) -> Result<()> {
+        if !cfg!(feature = "hotplug") {
+            return Err(VfioDeviceError::UpdateNotAllowedPostBoot);
+        }
+
+        slog::info!(
+            ctx.logger(),
+            "prepare remove VFIO device";
+            "subsystem" => "vfio_dev_mgr",
+            "hostdev_id" => hostdev_id,
+        );
+
+        let device_index = self
+            .get_index_of_hostdev_id(hostdev_id)
+            .ok_or(VfioDeviceError::InvalidConfig)?;
+
+        let info = &self.info_list[device_index];
+        if let Some(dev) = info.device.as_ref() {
+            let callback: Option<Box<dyn Fn(UpcallClientResponse) + Send>> =
+                Some(Box::new(move |result| match result {
+                    UpcallClientResponse::DevMgr(response) => {
+                        if let DevMgrResponse::Other(resp) = response {
+                            if let Err(e) = result_sender.send(Some(resp.result)) {
+                                error!("send upcall result failed, due to {:?}!", e);
+                            }
+                        }
+                    }
+                    UpcallClientResponse::UpcallReset => {
+                        if let Err(e) = result_sender.send(None) {
+                            error!("send upcall result failed, due to {:?}!", e);
+                        }
+                    }
+                    _ => {}
```

Comment:
Do something? Like print some logs or return an error?

Reply:
add the match arm for test case here.